### PR TITLE
feat(bench): add `InstructionTracer.js` helper

### PR DESCRIPTION
**Pull request by mooori**
Adds the `InstructionTracer.js` helper for `zkevm-proverjs` and integrates it with tooling by:

- Adding the `profile-instructions` command to `run-tests-zkasm.js`: It executes instrumented zkASM.
- Make codegen emit calls to the `InstructionTracer` helper when `ZkasmSettings::emit_profiling_info` is enabled.

Can be reviewed commit by commit.

## Usage

Generate instrumented zkASM with `analyze-zkasm`, for example:

```bash
cargo run -p analyze-zkasm -- instrument-inst \
    cranelift/zkasm_data/spectest/i32/add_1.wat \
    cranelift/zkasm_data/spectest/i32/generated/add_1.zkasm
```

Execute the instrumented zkASM and write the trace of executed instructions to a file:

```bash
node ./tests/zkasm/run-tests-zkasm.js profile-instructions \
    cranelift/zkasm_data/spectest/i32/generated/add_1.zkasm \
    trace.txt
```

## Testing

One of the next steps is adding a subcommand to `analyze-zkasm` that allows to go `wat file -> instrumented zkasm -> executed instruction trace` in one step. Adding tests for execution traces will be much easier once this command exists. Therefore, if possible, I would like to postpone adding tests until then.

### CODE REVIEW

File: cranelift/codegen/src/isa/zkasm/inst/emit.rs [hunk 1]
```diff
@@ -369,13 +369,16 @@ impl Inst {
...
+    /// Calls a method implemented by the `InstructionTracer.js` helper. An instance of this helper
+    /// must be provided to `zkevm-proverjs` for the successful execution of the generated zkASM.
     fn emit_inst_instrumentation(&self, sink: &mut MachBuffer<Inst>) {
         match self {
             // Labels are handled separately since benchmarking will provide a separate command to
             // analyze labels.
             &MInst::Label { .. } => {}
             _ => put_string(
-                ";; TODO(mooori) call the helper to trace `MInst` execution\n",
+                &format!("$${{traceInstruction({})}}\n", self.print_name()),
```
Comment (nagisa) [inline cranelift/codegen/src/isa/zkasm/inst/emit.rs:381]: Just a brain dump, nothing actionable for this PR:

IF we wanted this instrumentation to be performance-sensitive, instrumenting *just* around the labels would be the ideal thing to do. After all, the instructions executed in between any two labels can be recovered entirely by looking at the zkwasm code. Doing so would reduce the overhead of the instrumentation appreciably.

Building tooling like that sounds non-trivial however, nor do we care about instrumented code performance (do we?) so I think its perfectly fine to do what you've done.

File: tests/zkasm/helpers/InstructionTracer.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+     * @param {string} path 
+     */
+    writeRawTrace(path) {
+        if (typeof path !== "string") {
+            // Writing to a descriptor will append instead of replace content,
+            // which might result in invalid traces, see 
+            // https://nodejs.org/api/fs.html#using-fswritefile-with-file-descriptors
+            throw new Error("provide a file name (not descriptor) to write to");
+        }
+        fs.writeFileSync(path, this.rawTrace.join("\n"));
```
Comment (nagisa) [inline tests/zkasm/helpers/InstructionTracer.js:43]: How large do these traces get for real programs we want to look at? My intuition is that they are going to become pretty large after executing for not too long, so this code may very well be pretty limited by memory available on the machine. An easy improvement here would be to join and write out the instructions in chunks of, say, 1000 instructions or somesuch number. That would likely cut down the peak memory use by this in half with a reasonably little effort.

(Another future improvement that's more involved would be to collect into `this.rawTrace` numeric IDs of instructions into a `UInt8Array`/`Buffer` and write them out as a binary blob, that way a trace of 1M instructions would end up being just 1MB, and not a significant multiple of the number.)

Comment (nagisa) [reply tests/zkasm/helpers/InstructionTracer.js:43]: Though now that I think about it, we've a limit of 8M steps, don't we? So the trace is still in the realm of megabytes, which is most likely perfectly fine.

**Review by nagisa — APPROVED** [review]

File: cranelift/codegen/src/isa/zkasm/inst/emit.rs (hunk 1 above)
Comment (mooori) [reply cranelift/codegen/src/isa/zkasm/inst/emit.rs:381]: My understanding is that for the time being we don't really care about instrumented code performance and the programs that are profiled terminate relatively quickly (like the sha256 and Keccak benchmarks).

Based on the functionality added for instruction tracing it should be relatively simple to do the same for labels. I would be happy to implement it for labels if we have capacity or if it becomes neccessary in the future due to performance reasons.

**Pull request merged**